src/api/box-edit/BoxEdit.js

```diff
@@ -0,0 +1,200 @@
+}
+
+let BoxEditInstance = null;
+
+class BoxEdit {
+    client: ComServerClient;
+
+    extensionRequestQueue: Map<string, extensionRequestQueueItem>;
+
+    constructor() {
+        if (!(BoxEditInstance instanceof BoxEdit)) {
+            BoxEditInstance = this;
+        }
+        this.extensionRequestQueue = new Map();
+        return BoxEditInstance;
+    }
+
+    queueGetNativeAppNameFromLocal(extension: string): Promise<string> {
+        // There's already a pending or fulfilled request for the appname
+        if (this.extensionRequestQueue.has(extension)) {
+            const queueItem = this.extensionRequestQueue.get(extension);
+            if (!queueItem) {
+                throw new Error('Race condition re: queueGetNativeAppNameFromLocal');
+            }
+            return queueItem.promise;
+        }
+
+        const extensionRequest = {};
+        const appNameRequestPromise = new Promise((resolve, reject) => {
+            extensionRequest.resolve = resolve;
+            extensionRequest.reject = reject;
+        });
+        extensionRequest.promise = appNameRequestPromise;
+        this.extensionRequestQueue.set(extension, extensionRequest);
+
+        return appNameRequestPromise;
+    }
+
+    checkBoxEditAvailability(): Promise<any> {
+        return this.getBoxEditAvailability();
+    }
+
+    getBoxEditAvailability(): Promise<any> {
+        this.client = new ComServerClient(CONSTANTS.BOX_EDIT_APP_NAME);
+
+        return this.client.getComServerStatus();
+    }
+
+    async canOpenWithBoxEdit(extensions: string[]): Promise<any> {
+        const extensionToAppTuples: Array<stringTuple> = await Promise.all(
+            extensions.map(async ext => {
+                try {
+                    const appName = await this.getAppForExtension(ext);
+                    const result: stringTuple = [ext, appName];
+                    return result;
+                } catch (err) {
+                    const result: stringTuple = [ext, ''];
+                    return result;
+                }
+            }),
+        );
+
+        const resultMap: Map<string, string> = new Map();
+        extensionToAppTuples.forEach(tuple => resultMap.set(...tuple));
+
+        return Promise.resolve(resultMap);
+    }
+
+    openFile(fileTypedID: string, token: Object): Promise<any> {
+        // @NOTE. canOpenWithBoxEdit, create token taken care of higher levels
+        // therefore not ported into React library
+
+        // TODO is token the right name?
+        const executeDataAsString = createExecuteData(
+            fileTypedID.slice(2),
+            token.data.token,
+            token.data.auth_code,
+            token.data.token_scope,
+        );
+
+        return this.client.sendCommand(executeDataAsString, TIMEOUT_MS);
+    }
+
+    getAppForExtension(extension: string): Promise<any> {
+        try {
+            if (isBlacklistedExtension(extension)) {
+                throw new Error('blacklisted');
```

Comment:
If we want to record this in the future, could you also toss the extension that was blacklisted. ( I know we can check against the list of blacklisted extensions, but it'll allow us to see EXACTLY which ones are throwing this error)